### PR TITLE
Better `readInt` take two

### PR DESCRIPTION
**Pull request by vdukhovni**
This version still outperforms the original by ~4-5x, and now implements
correct overflow checks for the full range of `Int` values.

A "reasonable" number of leading zeros is supported after an optional
explicit sign, but once ~32k of leading zeros have been seen, no further
chunks will be processed and the read fails (returns `Nothing` as the
`Maybe Int` payload).

It also adds `skipSomeWS` which skips a "reasonable" quantity of
leading whitespace before giving up if more than ~32k of whitespace
is encountered.

The new `w8IsSpace` function is somewhat out of place here, but it
is more efficient than any stock versions I know of in any other
module (barring writing some FFI code for this).  It could be
kept internal, with folks who want to roll their own skipping of
leading whitespace also providing their own implementation of the
predicate...

[ EDIT: got rid of the `readIntSkip` function, turns out just providing `skipSomeWS` separately actually works just as well, with no need to provide the composition. ]

### CODE REVIEW

**Comment by vdukhovni** [comment]
Squashed the fixup (older Preludes don't have `(<>)` built-in). So the CI should now be green.

So please let me know which of #29 or #31 (this PR) you'd like to progress.

CC: @chessai, @hsyl20, @bodigrim, @sjakobi, @cartazio

**Comment by sjakobi** [comment]
@vdukhovni Would you mind pinging me less often about things related to `streaming-bytestring`? I'm currently not very interested in this package. Thanks! :)

**Comment by vdukhovni** [comment]
> @vdukhovni Would you mind pinging me less often about things related to `streaming-bytestring`? I'm currently not very interested in this package. Thanks! :)

Sure (sorry about a final notification this will cause).  The reason I thought this is relevant is that there's an open discussion re `readInt` wrt. `bytestring` and overflow detection, ... so I thought there'd be some relevance.  Apologies.  Will honour your request going forward...

File: Data/ByteString/Streaming/Char8.hs [hunk 1]
```diff
@@ -710,20 +711,215 @@ putStrLn :: MonadIO m => ByteString m r -> m r
...
-            Left r -> Compose (Just n :> return r)
-            Right (c,rest') -> if isDigit c
-               then Compose (Nothing :> (chunk bs >> cons' c rest'))
-               else Compose (Just n :> (chunk more >> cons' c rest'))
-        else return (Compose (Just n :> (chunk more >> rest)))
+-- | Bounds for Word# multiplication by 10 without overflow, and
+-- absolute values of Int bounds.
+intmaxWord, intminWord, intmaxQuot10, intmaxRem10, intminQuot10, intminRem10 :: Word
+intmaxWord = fromIntegral (maxBound :: Int)
+intminWord = fromIntegral (negate (minBound :: Int))
```
Comment (Bodigrim) [inline Data/ByteString/Streaming/Char8.hs:718]: `negate` is redundant ;)

File: Data/ByteString/Streaming/Char8.hs [hunk 2]
```diff
@@ -710,20 +711,215 @@ putStrLn :: MonadIO m => ByteString m r -> m r
...
+(intmaxQuot10, intmaxRem10) = intmaxWord `quotRem` 10
+(intminQuot10, intminRem10) = intminWord `quotRem` 10
+
+-- Predicate to test whether a 'Word8' value is either ASCII whitespace,
+-- or a unicode NBSP (U+00A0).  Optimised for ASCII text, with spaces
+-- as the most frequent whitespace characters.
+w8IsSpace :: Word8 -> Bool
+w8IsSpace = \ !w8 ->
```
Comment (Bodigrim) [inline Data/ByteString/Streaming/Char8.hs:726]: A static bitmask literal `Addr#` will be probably faster. Cf. https://github.com/Bodigrim/unicode-transforms/blob/master/Data/Unicode/Internal/Bits.hs

File: Data/ByteString/Streaming/Char8.hs [hunk 3]
```diff
@@ -710,20 +711,215 @@ putStrLn :: MonadIO m => ByteString m r -> m r
...
+
+-- Try to position the stream at the next non-whitespace input, by
+-- skipping leading whitespace.  Only a /reasonable/ quantity of
+-- whitespace will be skipped before giving up and returning the rest
+-- of the stream with any remaining whitespace.  Limiting the amount of
+-- whitespace consumed is a safety mechanism to avoid looping forever
+-- on a never-ending stream of whitespace from an untrusted source.
+-- For unconditional dropping of all leading whitespace, use `dropWhile`
+-- with a suitable predicate.
+skipSomeWS :: Monad m => ByteString m r -> ByteString m r
```
Comment (Bodigrim) [inline Data/ByteString/Streaming/Char8.hs:745]: I'm unfamiliar with `streaming-bytestring`, but do you need to export this function? Is it expected to be generally useful?

File: Data/ByteString/Streaming/Char8.hs [hunk 4]
```diff
@@ -710,20 +711,215 @@ putStrLn :: MonadIO m => ByteString m r -> m r
...
+                    ptr = p `plusPtr` (nbytes + signlen - 1)
+                go end ptr acc
+                mapM_ (poke p) signByte
+          where
+            signlen = if signByte == Nothing then 0 else 1
+
+            go :: Ptr Word8 -> Ptr Word8 -> Word -> IO ()
+            go end !ptr !_ | end == ptr = return ()
+            go end !ptr !a = do
+                let (q, r) = a `quotRem` 10
```
Comment (Bodigrim) [inline Data/ByteString/Streaming/Char8.hs:923]: GHC does not optimize division by a constant. I suggest to replace division by multiplication manually:

```haskell
-- | In 64-bit case
-- q = floor(floor((2^67+2)/10 * n) / 2^67)
--   = floor((2^67+2)/10 * n/2^67)
--   = floor(n/10 + (n/2^64 * 1/4)/10) = floor(n/10)
--     because n/2^64 * 1/4 < 1.
--
-- 32-bit case is the same, but uses 35 instead of 67.
quotRem10 :: Word -> (Word, Word)
quotRem21 n
    | finiteBitSize (0 :: Word) == 64
    = let q = (n `highMul` 14757395258967641293) `shiftR` 3 in
        (q, n - q `shiftL` 1 - q `shiftL` 3)
    | finiteBitSize (0 :: Word) == 32
    = let q = (n `highMul` 3435973837) `shiftR` 3 in
        (q, n - q `shiftL` 1 - q `shiftL` 3)
    | otherwise
    = n `quotRem` 10

highMul :: Word -> Word -> Word
highMul (W# x#) (W# y#) = W# high#
    where
        !(# high#, _ #) = timesWord2# x# y#
```

File: tests/Test.hs [hunk 5]
```diff
@@ -148,5 +318,6 @@ main = defaultMain $ testGroup "Tests"
     , testCase "groupBy: Char order" groupByCharOrder
     , testCase "findIndexOrEnd" goodFindIndex
     , testCase "Stream Interop" firstI
+    , testCase "readInt" readIntCases
```
Comment (Bodigrim) [inline tests/Test.hs:368]: Do you have a coverage report? It's difficult to check otherwise that all branches are tested.

Comment (fosskers) [reply tests/Test.hs:368]: Definitely not - the library overall is somewhat under-tested in terms of unit/property tests.

File: Data/ByteString/Streaming/Char8.hs (hunk 1 above)
Comment (vdukhovni) [reply Data/ByteString/Streaming/Char8.hs:718]: It wouldn't be redundant on a 1s-complement system, but perhaps none are supported by Haskell.
In any case, this is a constant, that's evaluated just once.  So I'm inclined to leave it in... Would that be OK?

File: Data/ByteString/Streaming/Char8.hs (hunk 2 above)
Comment (vdukhovni) [reply Data/ByteString/Streaming/Char8.hs:726]: Yeah, I was looking for that, but found it tricky to do right, thanks for the link, I'll compare...

File: Data/ByteString/Streaming/Char8.hs (hunk 1 above)
Comment (Bodigrim) [reply Data/ByteString/Streaming/Char8.hs:718]: Of course. It will be evaluated in compile time, I guess.

File: Data/ByteString/Streaming/Char8.hs (hunk 3 above)
Comment (vdukhovni) [reply Data/ByteString/Streaming/Char8.hs:745]: I think that skipping whitespace is generally useful, and limiting the quantity skipped makes it an interesting new element in the toolkit.  Perhaps @fosskers can comment on whether `skipSomeWS` makes sense to export or not. It is needed for bulk-reading a stream of whitespace-separated integer data, for example.

File: tests/Test.hs (hunk 5 above)
Comment (vdukhovni) [reply tests/Test.hs:368]: I don't have a coverage report. Any pointers on how to get one?

Comment (Bodigrim) [reply tests/Test.hs:368]: `stack test --coverage` and probably a similar incantation with Cabal.
`stack` is somewhat fragile in this aspect, you may have a better mileage with [`stack-2.5.0.1-rc`](https://github.com/commercialhaskell/stack/releases/tag/v2.5.0.1)

File: Data/ByteString/Streaming/Char8.hs (hunk 4 above)
Comment (vdukhovni) [reply Data/ByteString/Streaming/Char8.hs:923]: This is not a performance critical part of the code path, it is only used on overflow, to recover the initial digits of the input stream. I did not feel it merited optimising.  Also, will this actually be significantly faster?

Comment (Bodigrim) [reply Data/ByteString/Streaming/Char8.hs:923]: In my experience it is noticeably faster, but I agree that it may be not worth it for this codepath.

File: Data/ByteString/Streaming/Char8.hs (hunk 2 above)
Comment (vdukhovni) [reply Data/ByteString/Streaming/Char8.hs:726]: I tested the bitmask code, it is nice, but actually slightly slower.  With just 3 cases to test (SP, NBSP and [TAB..CR]), the carefully tuned `w8IsSpace` is faster than the bitmask lookup, because just computing the word index and offset and then doing the lookup is somewhat slower.  The runtime for reading 100M space separated ints goes up from 3.8s to 4.0s with the bitmask code...  If the set of characters to include were more complex this would definitely be the way to go, but not in this case...

[ I'd like to mark this one resolved if you agree ]

File: tests/Test.hs (hunk 5 above)
Comment (vdukhovni) [reply tests/Test.hs:368]: In terms of coverage, I made sure that overflow does not happen at maxBound/minBound, and does happen directly outside that range.  I also made sure that we're not performing the wrong number of I/O ops by rerunning already performed effects. And testing that leading zeros are handled correctly, and chunk boundaries are not an issue just after a sign, etc. and that the unconsumed tail of the string is correct.  Also tested that just too many leading zeros are rejected, and that overflow reconstruction is returning the original input.

It would be nice to know which branches are not hit if any, what tools expose that?

File: Data/ByteString/Streaming/Char8.hs (hunk 4 above)
Comment (vdukhovni) [reply Data/ByteString/Streaming/Char8.hs:923]: But I could do multiplication by `10` by shift and add, if that's better, unless GHC already does that...  The multiplications are on the critical path when accumulating digits...

File: tests/Test.hs (hunk 5 above)
Comment (Bodigrim) [reply tests/Test.hs:368]:
```
cabal test --enable-coverage
```
or 
```
stack upgrade --binary-version 2.5.0.1
stack test --coverage
```
Both options will produce a detailed HTML report, highlighting not hit branches.

File: Data/ByteString/Streaming/Char8.hs (hunk 2 above)
Comment (Bodigrim) [reply Data/ByteString/Streaming/Char8.hs:726]: Thanks, interesting. (For some reason I cannot resolve it myself)

File: tests/Test.hs (hunk 5 above)
Comment (vdukhovni) [reply tests/Test.hs:368]: Wow!  That was painless.  Got the code coverage to 100% (for the new code in this PR only) in just ~20 minutes. Just needed to add a couple of simple test cases, and replace a couple of variables with their constant value in that context. There is a cosmetic never evaluated `0` and `()` and two similar places for the accumulator `acc`, but those are [obviously fine](http://imrryr.org/~viktor/sbs-hpc/test/streaming-bytestring-0.1.6-inplace/Data.ByteString.Streaming.Char8.hs.html)

(Line 714 to end of file)

[ Edit I'm reverting the change that made the unevaluated `acc` argument of `overflow` be zero, because somehow doing that caused a noticeable performance regression, runtime jumped from 3.8s to 4.5s for the 100M bulk read!  Also tweaked the tests for overly large quotient to be precise, so that off-by one errors in the quotient are not in the future missed. ]

**Review by Bodigrim — APPROVED** [review]

File: Data/ByteString/Streaming/Char8.hs (hunk 3 above)
Comment (fosskers) [reply Data/ByteString/Streaming/Char8.hs:745]: Given that detecting whitespace is tricky to get right (see `w8IsSpace`), I think it's good for us to export `skipSomeWS`.

File: Data/ByteString/Streaming/Char8.hs [hunk 6]
```diff
@@ -710,20 +711,215 @@ putStrLn :: MonadIO m => ByteString m r -> m r
...
+    -- the conversion from Word8 to Word is free.
+    let w :: Word
+        !w = fromIntegral w8
+     in w - 0x21 > 0x7e   -- not [x21..0x9f]
+        && ( w == 0x20    -- SP
+          || w - 0x09 < 5 -- HT, NL, VT, FF, CR
+          || w == 0xa0 )  -- NBSP
+{-# INLINE w8IsSpace #-}
+
+-- Try to position the stream at the next non-whitespace input, by
```
Comment (fosskers) [inline Data/ByteString/Streaming/Char8.hs]:
```suggestion
-- | Try to position the stream at the next non-whitespace input, by
```

**Review by fosskers — CHANGES_REQUESTED** [review]

File: Data/ByteString/Streaming/Char8.hs (hunk 6 above)
Comment (vdukhovni) [reply Data/ByteString/Streaming/Char8.hs]: Thanks, pushed a fixup.  Once all the reviews are done, I'd like to squash all the commits before this is merged.

**Review by fosskers — APPROVED** [review]

**Comment by vdukhovni** [comment]
Many thanks!  Just pushed a squashed version.

**Comment by vdukhovni** [comment]
> Many thanks! Just pushed a squashed version.

It is the same as what you reviewed, just squashed into one commit.  So I think this is ready to be merged...

**Comment by vdukhovni** [comment]
Thanks!

**Comment by fosskers** [comment]
Prepping for release soon, once I get mine merged in too.
